pkg/transport/custom_transform_collection.go

```diff
@@ -0,0 +1,216 @@
+/*
+Copyright 2024 The KubeStellar Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package transport
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/klog/v2"
+
+	"github.com/kubestellar/kubestellar/api/control/v1alpha1"
+	controlclient "github.com/kubestellar/kubestellar/pkg/generated/clientset/versioned/typed/control/v1alpha1"
+	"github.com/kubestellar/kubestellar/pkg/jsonpath"
+)
+
+// customTransformCollection maintains the digested custom transformation instructions
+// for each metav1.GroupResource.
+type customTransformCollection struct {
+	// client is here for updating status in a CustomTransform
+	client controlclient.CustomTransformInterface
+
+	// getTransformObjects is the part of the CustomTransform informer's cache.Indexer behavior
+	// that is needed here, for using the index named in `customTransformDomainIndexName`.
+	getTransformObjects func(indexName, indexedValue string) ([]any, error)
+
+	// enqueue is used to add a reference to a Binding that needs to be re-processed because
+	// of a change to a CustomTransform that the Binding is sensitive to.
+	enqueue func(any)
+
+	// mutex must be locked while accessing the following fields or their contents
+	mutex sync.Mutex
```

Comment:
I think that the methods of customTransformCollection need their Lock and Unlock statements where they are, covering more than individual map actions. If the map actions _also_ tried to Lock then this would be bad.

Reply:
It looks to me like all the methods of customTransformCollection that need any lock, need a write lock.